rl/epstein_civil_violence/model.py

```diff
@@ -0,0 +1,167 @@
+        height=20,
+        citizen_density=0.5,
+        cop_density=0.05,
+        citizen_vision=4,
+        cop_vision=4,
+        legitimacy=0.8,
+        max_jail_term=30,
+        arrest_prob_constant=2.3,
+        movement=True,
+        max_iters=200,
+    ):
+        """
+        Initialize the EpsteinCivilViolenceRL environment.
+
+        Parameters:
+        - width: Width of the grid.
+        - height: Height of the grid.
+        - citizen_density: Density of citizens in the grid.
+        - cop_density: Density of cops in the grid.
+        - citizen_vision: Vision range of citizens.
+        - cop_vision: Vision range of cops.
+        - legitimacy: Legitimacy parameter of the model.
+        - max_jail_term: Maximum jail term for arrested citizens.
+        - arrest_prob_constant: Constant used in arrest probability calculation.
+        - movement: Flag indicating whether agents can move or not.
+        - max_iters: Maximum number of iterations for the model.
+        """
+
+        super().__init__(
+            width,
+            height,
+            citizen_density,
+            cop_density,
+            citizen_vision,
+            cop_vision,
+            legitimacy,
+            max_jail_term,
+            0,
+            arrest_prob_constant,
+            movement,
+            max_iters,
+        )
+
+        # Defining RL specific attributes
+        self.observation_space = gym.spaces.Box(
+            low=0, high=4, shape=(((cop_vision * 2 + 1) ** 2 - 1),), dtype=np.float32
+        )
+        self.action_space = gym.spaces.Tuple(
+            (gym.spaces.Discrete(8), gym.spaces.Discrete(5))
+        )
+
+    def step(self, action_dict):
+        """
+        Perform a step in the environment.
+
+        Parameters:
+        - action_dict: Dictionary containing actions for each agent.
+
+        Returns:
+        - observation: Current observation of the environment.
+        - rewards: Dictionary containing rewards for each agent.
+        - done: Dictionary indicating if each agent is done.
+        - truncated: Dictionary indicating if each agent is truncated.
+        - info: Additional information about the step.
+        """
+        # Update the action dictionary for step
+        self.action_dict = action_dict
+
+        # Step the model
+        self.schedule.step()
+        self.datacollector.collect(self)
+
+        # Calculate rewards
+        rewards = self.cal_reward()
+
+        # Update matrix for observation space
+        grid_to_observation(self, CitizenRL)
+        observation = {}
+        for agent in self.schedule.agents:
+            observation[agent.unique_id] = [
+                self.obs_grid[neighbor[0]][neighbor[1]]
+                for neighbor in agent.neighborhood
+            ]  # Get the values from the observation grid for the neighborhood cells
+
+        # RL specific outputs for the environment
+        done = {a.unique_id: False for a in self.schedule.agents}
```

Comment:
I see already a place where the brand new [AgentSet `set` method](https://github.com/projectmesa/mesa/pull/2254) can be useful!

Reply:
Is it working with Mesa 2?

Reply:
As of about three hours ago ;)

https://github.com/projectmesa/mesa/releases/tag/v2.3.3

Reply:
But you're creating a dict and not setting a variable, so I made a wrong suggestion.